[labs/router] Add initial implementation of labs/router (#2331)
lit/lit: packages/labs/router/README.md

- Extraction of URL pattern parameters into data objects passed to render callbacks
- A `routes.outlet` that renders the current route's render callback
- A `routes.link()` method to generate URLs to use in `<a>` tags, etc.
- A `routes.goto()` method for performing a navigation

## API

### Router

A Router is a controller (a subclass of the Routes controller) for use at the top-level of an application. It's main purpose is to set up global `click` and `popstate` event listeners (which should be installed only once on a page). It can optionally contain route definitions as a convenience.

It can be installed with no configuration:

```ts
class App extends LitElement {
  private router = new Router(this);
}
```

Or contain route configurations:

```ts
class MyElement extends LitElement {
  private router = new Router(this, [
    {path: '/', render: () => html`<h1>Home</h1>`},
  ]);

  render() {
    return router.outlet;
  }
}
```

### Routes

Routes is the main interface into the router API. A Routes controller contains route definitions and the templates that each route renders:

```ts
class MyElement extends LitElement {

>   private routes = new Routes(this, [

Reviewer: So this is Route*S* and not Route*R* as above?  Mistake?

Reviewer: I see you have Router inheriting from Routes. As the names look so similar, it might make sense to point out this difference or find a better name for one of the two

### Nested Routes

Nested routes allow child components to define a subset of the route space mounted at a URL prefix path chosen by a parent.

```ts
class XParent extends LitElement {
  private _routes = new Routes(this, [
    {path: 'foo', render: () => html`<x-foo></x-foo>`},
    // Here we mount a child component that defines its own sub-routes.
    // We need the trailing /* parameter to match on the prefix and pass
    // a path to the child to parse.
    {path: '/child/*', render: () => html`<x-child></x-child>`},
  ]);
  render() {
    return html`${this._routes.outlet}`;
  }
}

class XChild extends LitElement {
  private _routes = new Routes(this, [
    {path: 'foo', render: () => html`<x-foo></x-foo>`},
    {path: 'bar', render: () => html`<x-bar></x-bar>`},
  ]);
  render() {
    return html`${this._routes.outlet}`;
  }
}
```

In this example, the page can handle URLs `/foo`, `/child/foo` and `/child/bar`.

## TODO

### Server router integration

This client-side router is intended to be able to integrate with server-side routing, both configuration-based routers and convention-based routers like file-based routers.

> Current ideas for integration include a fallback route handler (instaled by the app) that delegates to a server API call that returns additional route configuration dynamically. The minimal API needed in the client router are:

Reviewer: installed (two l's)

A router for Lit.

## Status

`@lit-labs/router` is not yet published to npm.

This package requires either a native [`URLPattern`](https://developer.mozilla.org/en-US/docs/Web/API/URLPattern) implementation (which is currently only implemented in Chrome, Edge, and other Chromium browsers) or a URLPattern polyfill, like [`urlpattern-polyfill`](https://github.com/kenchris/urlpattern-polyfill).

## Overview

`@lit-labs/router` is a component-oriented router API vended as reactive controllers. Routes are configured as part of component definitions, and integrated into the component lifecycle and rendering.

Usage will generally look like this, with a configuration in a reactive controller, and rendering done via route-specific render callbacks and an "outlet" to use in the main render() method:

```ts
class MyElement extends LitElement {

>   private _routes = new Routes(this, [

Reviewer 2: The `this` in the render fns is still bound to the element here, so ``() => html`<h1>Home: ${this.someProp}</h1>` `` works, right?

>   private _routes = new Routes(this, [

Author: Yes, that's right.

>   private _routes = new Routes(this, [

Reviewer 3: Seems ok I guess.

    {path: '/', render: () => html`<h1>Home</h1>`},
    {path: '/projects', render: () => html`<h1>Projects</h1>`},
    {path: '/about', render: () => html`<h1>About</h3>`},
  ]);

  render() {
    return html`
      <header>...</header>

>       <main>${this._routes.outlet}</main>

Author: Yeah, I can make that change. I considered `this._routes.render()`, but I feel like we're getting a lot of `render()` methods that need to be differentiated. `renderOutlet()` is a bit longer, but more precise.

>       <main>${this._routes.outlet}</main>

Reviewer 3: What about just `outlet()` ?

>       <main>${this._routes.outlet}</main>

Author: done

      <footer>...</footer>
    `;
  }
}
```

Routes can be nested: a route path can include a trailing `/*` pattern to match against a prefix, and will automatically propagate that prefix to Routes controllers defined in child elements.

The general shape of the API includes:

- A `Router` controller that's used as a top-level singleton to set up event listeners
- A `Routes` controller for declaring routes inside components
- Declaration of routes with [`URLPattern`](https://developer.mozilla.org/en-US/docs/Web/API/URLPattern) and render callbacks
- Extraction of URL pattern parameters into data objects passed to render callbacks
- A `routes.outlet` that renders the current route's render callback
- A `routes.link()` method to generate URLs to use in `<a>` tags, etc.
- A `routes.goto()` method for performing a navigation

## API

### Router

> A Router is a controller (a subclass of the Routes controller) for use at the top-level of an application. It's main purpose is to set up global `click` and `popstate` event listeners (which should be installed only once on a page). It can optionally contain route definitions as a convenience.

Author: I had considered making `Router` _not_ extend `Routes` and just be a thing that installs the event listeners. For simple cases though a component would need to install both a `Router` and a `Routes`, and that seemed unfortunate.

React Router also has this Router/Routes division. In their case they have browser, server, and browser/fragment implementations. We're going to try to run the same router on the client and server, but one could imagine a Router implementation that uses the App History API instead of catching click and popstate events.

I'm hoping we can document that "Router goes with the app, Routes go with nested components" and that'll make sense.

> A Router is a controller (a subclass of the Routes controller) for use at the top-level of an application. It's main purpose is to set up global `click` and `popstate` event listeners (which should be installed only once on a page). It can optionally contain route definitions as a convenience.

Reviewer 3: Sounds good.

It can be installed with no configuration:

```ts
class App extends LitElement {
  private router = new Router(this);
}
```

Or contain route configurations:

```ts
class MyElement extends LitElement {
  private router = new Router(this, [
    {path: '/', render: () => html`<h1>Home</h1>`},
  ]);

  render() {
    return router.outlet;
  }
}
```

### Routes

Routes is the main interface into the router API. A Routes controller contains route definitions and the templates that each route renders:

```ts
class MyElement extends LitElement {
  private routes = new Routes(this, [
    {path: '/', render: () => html`<h1>Home</h1>`},
    {path: '/projects', render: () => html`<h1>Projects</h1>`},
    {path: '/about', render: () => html`<h1>About</h1>`},
  ]);
}
```

The second argument is the route configuration: an array of `RouteConfig` objects.

#### RouteConfig

A RouteConfig contains at the minimum the pattern to match URLs against and a template to render. Names can be provided to reference routes for link generation.

> There are two types of `RouteConfig`s: `PathRouteConfig` and `URLPatternRouteConfig`:

Author: This is there because if we make URLPattern objects for you, it seems good to let you bring your own. It's only the difference between:

```ts
{
  path: '/foo/:id',
  render({id}) { ... }
}
```

and

```ts
{
  pattern: new URLPattern({pathname: '/foo/:id'}),
  render({id}) { ... }
}
```

The key additional feature that making your own URLPattern allows now is to match on query params. That seems niche enough to relegate to bring-you-own-pattern-object.

> There are two types of `RouteConfig`s: `PathRouteConfig` and `URLPatternRouteConfig`:

Reviewer 3: OK.

}
```

#### Render callbacks

The render callback is called when the outlet property of the Routes object is accessed. It is passed an object with the parameters extracted from the matching URL.

Example with named parameter:

```ts
{
  path: '/profile/:id',
  render: ({id}) => html`<x-profile .profileId=${id}></x-profile>`
}
```

#### Outlets

An outlet is where a routes object renders the currently selected route's template. It can be used anywhere in the host element's template:

```ts
html`<main>${this.routes.outlet}</main>`;
```

#### enter() callbacks

A route can define an `enter()` callback that lets it do work before rendering and optionally reject that route as a match.

`enter()` can be used to load and wait for neccessary component definitions:

```ts
{
  path: '/*',
  render: (params) => html`<x-foo></x-foo>`,
  enter: async (params) => {
    await import('./x-foo.js');
  },
}
```

or dynamically install new routes:

```ts
{
  path: '/*',
  render: (params) => html`<h1>Not found: params[0]</h1>`,
  enter: async (params) => {
    const path = params[0];
    const dynamicRoute = getDynamicRoute(path);
    if (dynamicRoute) {
      const {routes} = this._router;

>       routes.splice(routes.langth - 1, 0, dynamicRoute);

Author: The splice is needed to insert before the fallback. We talked about making fallback a separate option so that this isn't needed.

> 

Reviewer 4: Maybe add a disclaimer that this isn't yet released?

Unless we're planning not to merge this to main until it's ready to release (I think I'd advocate for merging even before we're ready for a preview release, so that we keep the tests running with mainline dev).

## Overview

`@lit-labs/router` is a component-oriented router API vended as reactive controllers. Routes are configured as part of component definitions, and integrated into the component lifecycle and rendering.

Usage will generally look like this, with a configuration in a reactive controller, and rendering done via route-specific render callbacks and an "outlet" to use in the main render() method:

```ts
class MyElement extends LitElement {
  private _routes = new Routes(this, [
    {path: '/', render: () => html`<h1>Home</h1>`},
    {path: '/projects', render: () => html`<h1>Projects</h1>`},
    {path: '/about', render: () => html`<h1>About</h3>`},
  ]);

  render() {
    return html`
      <header>...</header>
      <main>${this._routes.outlet}</main>
      <footer>...</footer>
    `;
  }
}
```

Routes can be nested: a route path can include a trailing `/*` pattern to match against a prefix, and will automatically propagate that prefix to Routes controllers defined in child elements.

The general shape of the API includes:

- A `Router` controller that's used as a top-level singleton to set up event listeners
- A `Routes` controller for declaring routes inside components

> - Declaration of routes with `URLPatterns` and render callbacks

Reviewer 4: ```suggestion
- Declaration of routes with [`URLPattern`](https://developer.mozilla.org/en-US/docs/Web/API/URLPattern) and render callbacks
```

- A `routes.link()` method to generate URLs to use in `<a>` tags, etc.
- A `routes.goto()` method for performing a navigation

## API

### Router

A Router is a controller (a subclass of the Routes controller) for use at the top-level of an application. It's main purpose is to set up global `click` and `popstate` event listeners (which should be installed only once on a page). It can optionally contain route definitions as a convenience.

It can be installed with no configuration:

```ts
class App extends LitElement {
  private router = new Router(this);
}
```

Or contain route configurations:

```ts
class MyElement extends LitElement {
  private router = new Router(this, [
    {path: '/', render: () => html`<h1>Home</h1>`},
  ]);

  render() {
    return router.outlet;
  }
}
```

### Routes

Routes is the main interface into the router API. A Routes controller contains route definitions and the templates that each route renders:

```ts
class MyElement extends LitElement {
  private routes = new Routes(this, [
    {path: '/', render: () => html`<h1>Home</h1>`},
    {path: '/projects', render: () => html`<h1>Projects</h1>`},

>     {path: '/about', render: () => html`<h1>About</h3>`},

Reviewer 4: ```suggestion
    {path: '/about', render: () => html`<h1>About</h1>`},
```

  ]);
}
```

The second argument is the route configuration: an array of `RouteConfig` objects.

#### RouteConfig

A RouteConfig contains at the minimum the pattern to match URLs against and a template to render. Names can be provided to reference routes for link generation.

> There are two types of RoutConfig: `PathRouteConfig` and `URLPatternRouteConfig`:

Reviewer 4: ```suggestion
There are two types of `RouteConfig`s: `PathRouteConfig` and `URLPatternRouteConfig`:
```

`PathRouteConfig` lets you specify the URL pattern as a path string:

```ts
{name: 'home', path: '/', render: () => html`<h1>Home</h1>`}
```

```ts
export interface PathRouteConfig {
  name?: string | undefined;
  path: string;
  render: (params: {[key: string]: string}) => unknown;
}
```

`URLPatternRouteConfig` lets you specify the URL pattern as a [`URLPattern`](https://developer.mozilla.org/en-US/docs/Web/API/URLPattern) object:

```ts
{pattern: new URLPattern({pathname: '/'}), render: () => html`<h1>Home</h1>`}
```

```ts
export interface URLPatternRouteConfig {
  name?: string | undefined;
  pattern: URLPattern;
  render: (params: {[key: string]: string}) => unknown;
}
```

#### Render callbacks

The render callback is called when the outlet property of the Routes object is accessed. It is passed an object with the parameters extracted from the matching URL.

Example with named parameter:

```ts
{
  path: '/profile/:id',
  render: ({id}) => html`<x-profile .profileId=${id}></x-profile>`
}
```

#### Outlets

An outlet is where a routes object renders the currently selected route's template. It can be used anywhere in the host element's template:

```ts

> html`${this.routes.outlet}`;

Reviewer 4: ```suggestion
html`<main>${this.routes.outlet}</main>`;
```

Since you wouldn't usually need the `html` wrapper unless you also had some markup to wrap around it. Sticking in a `main` might be a common top-level wrapping pattern.

```

#### `goto()`

> `goto(url: string)` is a programmatic navigation API. It should be able to take full URLs for top-level navigation and relative URLs for navigation within a nested route space.

Reviewer 4: ```suggestion
`goto(url: string)` is a programmatic navigation API. It takes full URLs for top-level navigation and relative URLs for navigation within a nested route space.
```

> `goto(name: string, params: object)` _(not implemented)_ allows navigation via named routes. The name and params are scoped to the Routes object it's called on, though nested routes can be triggered by a "tail" parameter - the match of a trailing /\*parameter (See tail groups).

Reviewer 4: ```suggestion
`goto(name: string, params: object)` _(not implemented)_ allows navigation via named routes. The name and params are scoped to the Routes object it's called on, though nested routes can be triggered by a "tail" parameter - the match of a trailing `/*` parameter (See tail groups).
```

#### `link()`

Components need to generate links to resources within the app. It's desirable to not require that link generation has global knowledge of URLs, so links should be able to be generated with either only local information, or with abstracted parameters. `routes.link()` helps generate different kinds of links:

##### Relative links

Relative links are relative to the parent route and its current state, and can be specified with a path string or name and parameters.

Examples

> - `this._routes.link(user.id)` - within a `<x-user>` component that has a route pattern like ':id', this would link to another user profile.

Reviewer 4: ```suggestion
- `this._routes.link(user.id)` - within a `<x-user>` component that has a route pattern like `":id"`, this would link to another user profile.
```
